mautic/mautic
# BugFix - Segments that mix company and lead fields filters generate wrong SQL expressions

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Related user documentation PR URL | N/A
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/A
| BC breaks? | N/A
| Deprecations? | N/A

#### Description:
Right now, the way the segment filter queries are created is by separating the company filters and the lead filters and joining those expressions with AND. This approach prevents to combine company and lead filter expressions that use OR.
This PR takes all the segment filters in the same order as they are created in the UI and the resultant SQL expression matches that logic.

This PR also fixes the same issue as (https://github.com/mautic/mautic/pull/4188).

Note: All Mautic unit tests passed with this PR, as well as the api-library unit tests

#### Steps to reproduce the bug:
1. Create a new segment
2. Add a new lead field with filter `Email equals 'a'`
3. Add a new company field with filter `Company Email equals 'a'` and select OR as the boolean operator.

![image](https://user-images.githubusercontent.com/2924026/27942611-71073c1a-6295-11e7-87cf-df34f9aa29f1.png)

4. Save the segment
5. Take note of the new segment ID
6. On the console, execute: `app/console m:s:u -vvv -e dev -i SEGMENT_ID` (replace SEGMENT_ID with the new segment ID)
7. Note the SQL query that was generated, it has the following form:
<pre>
SELECT count(distinct(l.id)) as lead_count, max(l.id) as max_id 
FROM leads l INNER JOIN lead_lists_leads ll ON l.id = ll.lead_id 
WHERE (ll.date_added <= '2017-07-07 03:29:15') AND (ll.manually_added = :false) AND (ll.leadlist_id = 5) AND (l.id NOT IN (SELECT l.id FROM leads l INNER JOIN companies_leads cl ON l.id = cl.lead_id INNER JOIN companies comp ON cl.company_id = comp.id 
WHERE (comp.companyemail = :fkWBKwIU) <b>AND</b> (l.email = :ZlaInqsc) GROUP BY l.id))
</pre>

#### Steps to test this PR:
1. Apply this PR
2. Execute the same command in the console: `app/console m:s:u -vvv -e dev -i SEGMENT_ID`
3. Note the SQL query that was generated, this time it will be correct:
<pre>
SELECT count(distinct(l.id)) as lead_count, max(l.id) as max_id 
FROM leads l INNER JOIN lead_lists_leads ll ON l.id = ll.lead_id 
WHERE (ll.date_added <= '2017-07-07 03:29:15') AND (ll.manually_added = :false) AND (ll.leadlist_id = 5) AND (l.id NOT IN (SELECT l.id FROM leads l INNER JOIN companies_leads cl ON l.id = cl.lead_id INNER JOIN companies comp ON cl.company_id = comp.id 
WHERE (comp.companyemail = :fkWBKwIU) <b>OR</b> (l.email = :ZlaInqsc) GROUP BY l.id))
</pre>

(A more thourough test would include a combination of several lead and company fields joined by AND or OR. The resultant SQL should also match the logic.)
